crypto/stark-verifier-ethereum/contracts/utils.sol

```diff
@@ -0,0 +1,68 @@
+pragma solidity ^0.6.4;
+
+
+library Utils {
+    function bit_reverse(uint64 num, uint8 number_of_bits) internal view returns (uint256 num_reversed) {
+        uint64 n = num;
+        uint64 r = 0;
+        for (uint8 k = 0; k < number_of_bits; k++) {
+            r = (r * 2) | (n % 2);
+            n = n / 2;
+        }
+        return r;
+    }
+
+    function bits_in(uint64 data) internal pure returns (uint8) {
+        uint8 result = 0;
+        if (data >= (1 << 32)) {
+            result += 32;
+            data >>= 32;
+        }
+        if (data >= (1 << 16)) {
+            result += 16;
+            data >>= 16;
+        }
+        if (data >= (1 << 8)) {
+            result += 8;
+            data >>= 8;
+        }
+        if (data >= (1 << 4)) {
+            result += 4;
+            data >>= 4;
+        }
+        if (data >= (1 << 2)) {
+            result += 2;
+            data >>= 2;
+        }
+        if (data >= 2) {
+            result += 1;
+            data >>= 1;
+        }
+        return result + uint8(data) - 1;
+    }
+
+    function deep_copy(bytes32[] memory a, bytes32[] memory b) internal pure {
+        for (uint256 i = 0; i < a.length; i++) {
```

Comment:
`i` can just be of type `u32`?

Reply:
No because then you need to cast it to index the array.

Reply:
@z2trillion  The default type in Solidity is `u256` and other types cause more problems. Best to avoid them if you can.